### PR TITLE
[bcl][jit] implement Interlocked.Exchange<T> in terms of object

### PR DESCRIPTION
**Pull request by lambdageek**
`Exchange<T>` is supposed to be picked up as an intrinsic.  But if it isn't (until #17338 is merged),
since `T` has a class constraint, call the `Exchange (ref object, ref object, ref object)` overload instead.

This works around a limitation in the JIT:
https://github.com/mono/mono/blob/82a07273c22b996b08fa88ee2e6632d782ac2242/mono/mini/mini-trampolines.c#L704-L713

if we're in the common call trampoline, and the caller is generic method and the callee is a generic method, if we're using generic sharing for the caller, but there's no generic jit info for the callee, we will not patch the call site.

In this case we had `Exchange<T>` calling `Exchange_T<T>`
where the callee is an icall (and evidently didn't have generic jit info).

So every time we called `Exchange<T>`, we would go through the trampoline when trying to call `Exchange_T<T>` without patching the call site.

Addresses part of https://github.com/mono/mono/issues/17334

### CODE REVIEW

**Comment by lambdageek** [comment]
H/T to @lewurm who did the hard work of identifying the cause.

**Comment by lambdageek** [comment]
@monojenkins build deb with monolite

**Comment by lambdageek** [comment]
I should also get rid of `CompareExchange_T<T>` for the same reason, actually

**Comment by lambdageek** [comment]
@monojenkins build deb with monolite

**Comment by vargaz** [comment]
Not sure this will work:
src/System.Threading/Interlocked.cs(94,25): error CS1503: Argument 1: cannot convert from 'ref T' to 'ref object' [/Users/vargaz/git/netcore/netcore/System.Private.CoreLib/System.Private.CoreLib.csproj]

**Comment by vargaz** [comment]
It might need some Unsafe method usage.

**Comment by jaykrell** [comment]
Did I break it when I did coop conversion?

**Comment by jaykrell** [comment]
I don't think so, since I thought I preserved the layering. Only change should be, when it wasn't inlined/intrinsic. Let me double check..

**Comment by lambdageek** [comment]
Yea this doesn't work.
@vargaz I don't think there's an Unsafe method that does this.

@jaykrell the issue is that there's an extra generic method->generic icall callsite now.  And it's hitting a condition where the trampoline code is deciding whether it can patch that callsite.  Separately, it happens that `Exchange<T>` is an intrinsic that isn't being intrinsifies correctly, and that's exposing this second issue.

**Comment by vargaz** [comment]
I.e.:
			CompareExchange (ref Unsafe.As<T, object> (ref location1), ref Unsafe.As<T, object> (ref value), ref Unsafe.As<T, object> (ref comparand), ref Unsafe.As<T, object> (ref result));

**Comment by lambdageek** [comment]
@vargaz thanks!

**Comment by vargaz** [comment]
Should probably add some [Intrinsic] attributes as well.

**Comment by jaykrell** [comment]
- So, it wasn't being intrinsified. Not my fault. A bit slow.
 - But then, when I changed it, I also broke patching, making it much slower.
 - Both are fixable.

File: mono/mini/aot-runtime.c [hunk 1]
```diff
@@ -4810,34 +4810,14 @@ mono_aot_get_method (MonoDomain *domain, MonoMethod *method, MonoError *error)
 		gboolean volatil = FALSE;
 		MonoMethodSignature *sig;
 
-		/* Same for CompareExchange<T> and Exchange<T> */
 		/* Same for Volatile.Read<T>/Write<T> */
```
Comment (jaykrell) [inline mono/mini/aot-runtime.c]: Maybe similar to do here?

Comment (lambdageek) [reply mono/mini/aot-runtime.c]: Oh yea that looks like dead code.  We already have method bodies like this.

In framework Mono:
https://github.com/mono/mono/blob/73ae9dbc3693201939efce02b2b36c4ce5b352bd/mcs/class/corlib/System.Threading/Volatile.cs#L166-L172

and in netcore Mono:
https://github.com/mono/mono/blob/73ae9dbc3693201939efce02b2b36c4ce5b352bd/netcore/System.Private.CoreLib/shared/System/Threading/Volatile.cs#L220-L229

**Comment by lambdageek** [comment]
> Does comparand need it?

Could do:
```csharp
   object v = value;
   object c = comparand;
   Exchange (...., ref v, ref c, ...);
```

**Comment by lambdageek** [comment]
@monojenkins build deb with monolite

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by lambdageek** [comment]
@monojenkins build

**Comment by vargaz** [comment]
The netcore build is missing a using:
https://dev.azure.com/dnceng/public/_build/results?buildId=388963

File: mcs/class/corlib/System.Threading/Interlocked.cs [hunk 2]
```diff
@@ -140,7 +141,8 @@ public static object Exchange (ref object location1, object value)
 			// This is not entirely convincing due to lack of volatile.
 			//
 			T result = null;
-			CompareExchange_T (ref location1, ref value, ref comparand, ref result);
+			// T : class so call the object overload.
+			CompareExchange (ref Unsafe.As<T, object> (ref location1), ref Unsafe.As<T, object>(ref value), ref Unsafe.As<T, object>(ref comparand), ref Unsafe.As<T, object>(ref result));
```
Comment (jkotas) [inline mcs/class/corlib/System.Threading/Interlocked.cs]:
```suggestion
			CompareExchange (ref Unsafe.As<T, object> (ref location1), value, comparand, ref Unsafe.As<T, object>(ref result));
```

File: netcore/System.Private.CoreLib/src/System.Threading/Interlocked.cs [hunk 3]
```diff
@@ -75,10 +75,6 @@ public static object Exchange (ref object location1, object value)
 		[MethodImplAttribute(MethodImplOptions.InternalCall)]
 		public extern static double CompareExchange (ref double location1, double value, double comparand);
 
-		[MethodImplAttribute(MethodImplOptions.InternalCall)]
-		[return: NotNullIfNotNull("location1")]
-		extern static void CompareExchange_T<T> (ref T location1, ref T value, ref T comparand, ref T result) where T : class?;
-
 		[return: NotNullIfNotNull("location1")]
 		public static T CompareExchange<T> (ref T location1, T value, T comparand) where T : class?
```
Comment (filipnavara) [inline netcore/System.Private.CoreLib/src/System.Threading/Interlocked.cs]: `[Intrinsic]` here as well?

**Comment by jaykrell** [comment]
netcore errors.

```
  Restore completed in 337.08 ms for /home/vsts/work/1/s/netcore/System.Private.CoreLib/System.Private.CoreLib.csproj.
/home/vsts/.nuget/packages/microsoft.sourcelink.common/1.0.0-beta2-19367-01/build/Microsoft.SourceLink.Common.targets(50,5): warning : Source control information is not available - the generated source link is empty. [/home/vsts/work/1/s/netcore/System.Private.CoreLib/System.Private.CoreLib.csproj]
src/System.Threading/Interlocked.cs(94,25): error CS0103: The name 'Unsafe' does not exist in the current context [/home/vsts/work/1/s/netcore/System.Private.CoreLib/System.Private.CoreLib.csproj]
src/System.Threading/Interlocked.cs(94,68): error CS0103: The name 'Unsafe' does not exist in the current context [/home/vsts/work/1/s/netcore/System.Private.CoreLib/System.Private.CoreLib.csproj]
src/System.Threading/Interlocked.cs(94,106): error CS0103: The name 'Unsafe' does not exist in the current context [/home/vsts/work/1/s/netcore/System.Private.CoreLib/System.Private.CoreLib.csproj]
src/System.Threading/Interlocked.cs(94,148): error CS0103: The name 'Unsafe' does not exist in the current context [/home/vsts/work/1/s/netcore/System.Private.CoreLib/System.Private.CoreLib.csproj]
src/System.Threading/Interlocked.cs(119,18): error CS0103: The name 'Unsafe' does not exist in the current context [/home/vsts/work/1/s/netcore/System.Private.CoreLib/System.Private.CoreLib.csproj]
src/System.Threading/Interlocked.cs(119,59): error CS0103: The name 'Unsafe' does not exist in the current context [/home/vsts/work/1/s/netcore/System.Private.CoreLib/System.Private.CoreLib.csproj]
src/System.Threading/Interlocked.cs(119,97): error CS0103: The name 'Unsafe' does not exist in the current context [/home/vsts/work/1/s/netcore/System.Private.CoreLib/System.Private.CoreLib.csproj]

Build FAILED.

/home/vsts/.nuget/packages/microsoft.sourcelink.common/1.0.0-beta2-19367-01/build/Microsoft.SourceLink.Common.targets(50,5): warning : Source control information is not available - the generated source link is empty. [/home/vsts/work/1/s/netcore/System.Private.CoreLib/System.Private.CoreLib.csproj]
src/System.Threading/Interlocked.cs(94,25): error CS0103: The name 'Unsafe' does not exist in the current context [/home/vsts/work/1/s/netcore/System.Private.CoreLib/System.Private.CoreLib.csproj]
src/System.Threading/Interlocked.cs(94,68): error CS0103: The name 'Unsafe' does not exist in the current context [/home/vsts/work/1/s/netcore/System.Private.CoreLib/System.Private.CoreLib.csproj]
src/System.Threading/Interlocked.cs(94,106): error CS0103: The name 'Unsafe' does not exist in the current context [/home/vsts/work/1/s/netcore/System.Private.CoreLib/System.Private.CoreLib.csproj]
src/System.Threading/Interlocked.cs(94,148): error CS0103: The name 'Unsafe' does not exist in the current context [/home/vsts/work/1/s/netcore/System.Private.CoreLib/System.Private.CoreLib.csproj]
src/System.Threading/Interlocked.cs(119,18): error CS0103: The name 'Unsafe' does not exist in the current context [/home/vsts/work/1/s/netcore/System.Private.CoreLib/System.Private.CoreLib.csproj]
src/System.Threading/Interlocked.cs(119,59): error CS0103: The name 'Unsafe' does not exist in the current context [/home/vsts/work/1/s/netcore/System.Private.CoreLib/System.Private.CoreLib.csproj]
src/System.Threading/Interlocked.cs(119,97): error CS0103: The name 'Unsafe' does not exist in the current context [/home/vsts/work/1/s/netcore/System.Private.CoreLib/System.Private.CoreLib.csproj]
    1 Warning(s)
```

**Comment by migueldeicaza** [comment]
From Nathan, with this fix and this one [1] Mono/LLVM/.NET5 configuration stands like this:

```
Mono/LLVM/.NET5 before change (req/sec):  1,478,095.55
Mono/LLVM/.NET5 after chang (req/sec) :   5,414,536.91
.NET Core 3/RyuJIT (req/sec):             6,626,015.90
```

So about 80% of the performance of .NET Core 3 now, looking good!

[1] https://github.com/mono/mono/pull/17338

**Pull request merged**